core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/ProgressImpl.java

```diff
@@ -0,0 +1,151 @@
+import javax.persistence.TemporalType;
+
+import com.eaglegenomics.simlims.core.User;
+
+import uk.ac.bbsrc.tgac.miso.core.data.impl.UserImpl;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.AbstractProgressStep;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.ProgressStep;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.Workflow.WorkflowType;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.Progress;
+
+@Entity
+@Table(name = "WorkflowProgress")
+public class ProgressImpl implements Progress {
+  @Id
+  @GeneratedValue(strategy = GenerationType.AUTO)
+  private long workflowProgressId;
+
+  @Enumerated(EnumType.STRING)
+  @Column(name = "workflowName")
+  private WorkflowType workflowType;
+
+  @ManyToOne(targetEntity = UserImpl.class)
+  @JoinColumn(name = "userId", nullable = false, updatable = false)
+  private User user;
+
+  @Temporal(TemporalType.TIMESTAMP)
+  @Column(name = "created", nullable = false, updatable = false)
+  private Date creationTime;
+
+  @Temporal(TemporalType.TIMESTAMP)
+  @Column(name = "lastModified", nullable = false)
+  private Date lastModified;
+
+  @OneToMany(targetEntity = AbstractProgressStep.class, mappedBy = "id.progress")
+  private List<ProgressStep> steps;
+
+  @Override
+  public long getId() {
+    return workflowProgressId;
+  }
+
+  @Override
+  public void setId(long id) {
+    this.workflowProgressId = id;
+  }
+
+  @Override
+  public WorkflowType getWorkflowType() {
+    return workflowType;
+  }
+
+  @Override
+  public void setWorkflowType(WorkflowType workflowType) {
+    this.workflowType = workflowType;
+  }
+
+  @Override
+  public User getUser() {
+    return user;
+  }
+
+  @Override
+  public void setUser(User user) {
+    this.user = user;
+  }
+
+  @Override
+  public Date getCreationTime() {
+    return creationTime;
+  }
+
+  @Override
+  public void setCreationTime(Date creationTime) {
+    this.creationTime = creationTime;
+  }
+
+  @Override
+  public Date getLastModified() {
+    return lastModified;
+  }
+
+  @Override
+  public void setLastModified(Date lastModified) {
+    this.lastModified = lastModified;
+  }
+
+  @Override
+  public List<ProgressStep> getSteps() {
+    return steps;
+  }
+
+  @Override
+  public void setSteps(List<ProgressStep> steps) {
+    Collections.sort(steps);
+    this.steps = steps;
+  }
+
+  @Override
+  public int hashCode() {
+    final int prime = 31;
+    int result = 1;
+    result = prime * result + ((creationTime == null) ? 0 : creationTime.hashCode());
+    result = prime * result + ((lastModified == null) ? 0 : lastModified.hashCode());
+    result = prime * result + ((steps == null) ? 0 : steps.hashCode());
+    result = prime * result + ((user == null) ? 0 : user.hashCode());
+    result = prime * result + (int) (workflowProgressId ^ (workflowProgressId >>> 32));
+    result = prime * result + ((workflowType == null) ? 0 : workflowType.hashCode());
+    return result;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 21 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)